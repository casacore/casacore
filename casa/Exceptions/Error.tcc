//# Error.cc: Base class for all AIPS++ errors (templated classes)
//# Copyright (C) 1993,1994,1995,1996,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/Exceptions/Error.h>
#include <casa/iostream.h>


namespace casa { //# NAMESPACE CASA - BEGIN

template<class t>
indexError<t>::indexError( t oI, const Char *str, Category c )
: IndexError(str,c),
  oIndex    (oI)
{}

template<class t>
indexError<t>::indexError( t oI, const String &str, Category c )
: IndexError(str,c),
  oIndex    (oI)
{}

template<class t>
indexError<t>::~indexError() throw()
{}


template<class t>
duplError<t>::duplError(t oI, const Char *str,Category c)
: DuplError(str,c),
  oKey     (oI)
{}

template<class t>
duplError<t>::duplError(t oI, const String &str,Category c)
: DuplError(str,c),
  oKey     (oI)
{}

template<class t>
duplError<t>::~duplError() throw()
{}


} //# NAMESPACE CASA - END

