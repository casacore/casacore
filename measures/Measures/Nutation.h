//# Nutation.h: Nutation class
//# Copyright (C) 1995,1996,1997,1998,2003,2004
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$

#ifndef MEASURES_NUTATION_H
#define MEASURES_NUTATION_H

//# Includes
#include <casacore/casa/aips.h>
#include <casacore/casa/Quanta/Quantum.h>
#include <casacore/casa/Quanta/Euler.h>

namespace casa { //# NAMESPACE CASA - BEGIN

//# Forward Declarations

// <summary> Nutation class and calculations </summary>

// <use visibility=export>

// <reviewed reviewer="Tim Cornwell" date="1996/07/01" tests="tMeasMath"
//	demos="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class=Measure>Measure</linkto> class
//   <li> <linkto class=Euler>Euler</linkto>
//   <li> <linkto class=MeasData>MeasData</linkto> class for constants
// </prerequisite>
//
// <etymology>
// Nutation
// </etymology>
//
// <synopsis>
// Nutation forms the class for Nutation calculations. It is a simple
// container with the selected method, and the mean epoch.
// It acts as a cache
// for values and their derivatives, to enable fast calculations for time
// epochs close together (see the <em>aipsrc</em> variable
// <em>measures.nutation.d_interval</em>).
//
// The calculation method is selected from one of the following:
// <ul>
//   <li> Nutation::STANDARD   (at 1995/09/04 the IAU1980 definition,
//				from 2004/01/01 IAU2000B)
//   <li> Nutation::NONE       (nutation of zero returned)
//   <li> Nutation::IAU1980
//   <li> Nutation::B1950
//   <li> Nutation::IAU2000
//   <li> Nutation::IAU2000A   (equal to the full precision (uas) IAU2000)
//   <li> Nutation::IAU2000B   (official lower precision (mas) of IAU2000)
// </ul>
// Epochs can be specified as the MJD (with defined constants MeasData::MJD2000
// and MeasData::MJDB1950 or the actual MJD),
// leading to the following constructors:
// <ul>
//   <li> Nutation() default; assuming STANDARD
//   <li> Nutation(method) 
// </ul>
// Actual Nutation for a certain Epoch is calculated by the () operator
// as Nutation(epoch), with epoch Double MJD. Values are returned as an
// <linkto class=Euler>Euler</linkto>.
// The derivative (d<sup>-1</sup>) can be obtained as well by 
// derivative(epoch). <br>
// A Nutation can be re-initialed with a different method and/or zero
// epoch with the <src>init()</src> functions (same format as constructors).
// To bypass the full calculation actual returned values are calculated
// using the derivative if within about 2 hours (error less than about
// 10<sup>-5</sup> mas). A call to refresh() will re-initiate calculations
// from scratch.<br>
// The following details can be set with the 
// <linkto class=Aipsrc>Aipsrc</linkto> mechanism:
// <ul>
//  <li> measures.nutation.d_interval: approximation interval as time 
//	(fraction of days is default unit) over which linear approximation
//	is used (default 0.04d (about 1 hour)).
//  <li> measures.nutation.b_usejpl: use the JPL database nutations for
//		 IAU1980.
//	Else analytical expression, relative error about 10<sup>-9</sup>
//	Note that the JPL database to be used can be set with 
//		measures.jpl.ephemeris (at the moment of writing DE200
//		 (default), or DE405)
//  <li> measures.nutation.b_useiers: use the IERS Database nutation
//		 corrections for IAU1980 (default False)
// </ul>
// </synopsis>
//
// <example>
//  <srcblock>
// #include <casacore/measures/Measures.h>
//	MVDirection pos(Quantity(10,"degree"),Quantity(-10.5,"degree"));
//						// direction RA=10; DEC=-10.5
//	Nutation mine(Nutation::IAU1980);	// define nutation type
//	RotMatrix rotat(mine(45837.0));		// rotation matrix for 84/05/17
//	MVDirection new = rotat*pos;		// apply nutation
//  </srcblock>
// The normal way to use Nutation is by using the
// <linkto class=MeasConvert>MeasConvert</linkto> class.
// </example>
//
// <motivation>
// To calculate the Nutation angles. An alternate route could have been
// a global function, but having a simple container allows caching of some
// calculations for speed.<br>
// Using MJD (JD-2400000.5) rather than JD is for precision reasons.
// </motivation>
//
// <todo asof="2003/10/20">
//   <li> Correct deval_p (derivative complimentary eqox terms)
//   <li> Improve speed by a bit more lazyness in derivative calculations
//	and separate eqox calculations 
// </todo>

class Nutation {
 public:
  //# Constants
  // Interval to be used for linear approximation (in days)
  static const Double INTV;
  
  //# Enumerations
  // Types of known Nutation calculations (at 1995/09/04 STANDARD == IAU1980,
  //	after 2004/01/01 it will be IAU2000B))
  enum NutationTypes {
    NONE, IAU1980, B1950, IAU2000A, IAU2000B,
    IAU2000 = IAU2000A,
    STANDARD = IAU1980 };
  
  //# Constructors
  // Default constructor, generates default J2000 Nutation identification
  Nutation();
  // Copy constructor
  Nutation(const Nutation &other);
  // Constructor with type
  explicit Nutation(NutationTypes type);
  // Copy assignment
  Nutation &operator=(const Nutation &other);
  
  //# Destructor
  ~Nutation();
  
  //# Operators
  // Return the Nutation angles
  const Euler &operator()(Double epoch);
  
  //# General Member Functions
  // Return derivative of Nutation (d<sup>-1</sup>)
  const Euler &derivative(Double epoch);
  
  // Re-initialise Nutation object
  // <group>
  void init();
  void init(NutationTypes type);
  // </group>
  
  // Refresh calculations
  void refresh();
  
  // Get the equation of equinox
  // <group>
  Double eqox(Double epoch) ;
  Quantity getEqoxAngle(Double epoch);
  Quantity getEqoxAngle(Double epoch, const Unit &unit) ;
  // </group>
  // Get the derivative of the equation of equinoxes in d<sup>-1</sup>
  Double derivativeEqox(Double epoch);
  // Get the complimentary terms of the equation of equinoxes
  Double eqoxCT(Double epoch);
  // Get the derivative of the complimentary terms of the equation of equinoxes
  Double derivativeEqoxCT(Double epoch);

 private:
  
  //# Data members
  // Method to be used
  NutationTypes method_p;
  // Check epoch for linear approximation
  Double checkEpoch_p;
  // Check epoch for calculation of derivatives
  Double checkDerEpoch_p;
  // Cached calculated angles
  Double nval_p[3];
  // Cached derivatives
  Double dval_p[3];
  // Cached equation of equinoxes
  Double eqeq_p;
  // Cached derivative equation of equinoxes
  Double deqeq_p;
  // Cached complimentary terms equation of equinoxes
  Double neval_p;
  // Cached derivative of complimentary terms equation of equinoxes
  Double deval_p;
  // To be able to use references rather than copies, and also to use these
  // references in simple (up to 4 terms of Nutation results) expressions,
  // results are calculated in circulating buffer
  Int lres_p;
  // Last calculation
  Euler result_p[4];
  // Interpolation interval
  static uInt myInterval_reg;
  // IERS use
  static uInt myUseiers_reg;
  // JPL use
  static uInt myUsejpl_reg;
  //# Member functions
  // Make a copy
  void copy(const Nutation &other);
  // Fill an empty copy
  void fill();
  // Calculate Nutation angles for time t; also derivatives if True given
  void calcNut(Double t, Bool calcDer = False);
};


} //# NAMESPACE CASA - END

#endif


