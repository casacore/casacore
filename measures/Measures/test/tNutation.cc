//# tNutation.cc:  Test program for parallel Nutation calculation
//# Copyright (C) 2013
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: Nutation.cc 21420 2014-03-19 09:18:51Z gervandiepen $

//# Includes
#include <measures/Measures/Nutation.h>
#include <casa/Exceptions/Error.h>

using namespace casa;

void doIt (int nthr, int n)
{
  double incr = 1./nthr;
  // Do nutation a number of times, if possible in parallel.
#ifdef _OPENMP
#pragma omp parallel for
#endif
  for (int i=0; i<nthr; ++i) {
    Double dat = 51116 + i*incr;
    ostringstream ostr;
    ostr << i;
    ofstream os(("tNutation_tmp.out_a" + ostr.str()).c_str());
    Nutation nut;
    for (int j=0; j<n; ++j) {
      os << dat << " Euler=" << nut(dat) << endl;
      os << " derivative=" << nut.derivative(dat) << endl;
      dat += incr;
    }
  }
}

int main(int argc, char* argv[])
{
  int nthr = 4;
  int n = 32;
  if (argc > 1) nthr = atoi(argv[1]);
  if (argc > 2) n    = atoi(argv[2]);
  try {
    doIt (nthr, n);
  } catch (const std::exception& x) {
    cout << "Unexpected exception: " << x.what() << endl;
    return 1;
  }
  return 0;
}
