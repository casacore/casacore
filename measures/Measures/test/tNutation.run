#!/bin/sh

rm -rf tNutation_tmp.out_*
# For time being use 1 thread until build with Boost is always done.
# Otherwise CountedPtr may not be thread-safe.
OMP_NUM_THREADS=1
export OMP_NUM_THREADS
$casa_checktool ./tNutation 4 32
for nm in a0 a1 a2 a3
do
  echo ''
  echo " *** tNutation_tmp.out_$nm"
  cat tNutation_tmp.out_$nm
done
