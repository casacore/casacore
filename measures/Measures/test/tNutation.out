
 *** tNutation_tmp.out_a0
51116 Euler=[0.409095, 5.17425e-05, -0.409058]
 derivative=[-6.21396e-09, 2.06725e-07, -1.19727e-07]
51116.2 Euler=[0.409095, 5.18002e-05, -0.409058]
 derivative=[-6.21396e-09, 2.54889e-07, -1.15973e-07]
51116.5 Euler=[0.409095, 5.18699e-05, -0.409058]
 derivative=[-6.21396e-09, 3.02605e-07, -1.0927e-07]
51116.8 Euler=[0.409095, 5.19514e-05, -0.409059]
 derivative=[-6.21396e-09, 3.48999e-07, -9.95502e-08]
51117 Euler=[0.409095, 5.20442e-05, -0.409059]
 derivative=[-6.21396e-09, 3.93136e-07, -8.68045e-08]
51117.2 Euler=[0.409095, 5.21477e-05, -0.409059]
 derivative=[-6.21396e-09, 4.34032e-07, -7.10869e-08]
51117.5 Euler=[0.409095, 5.22609e-05, -0.409059]
 derivative=[-6.21396e-09, 4.70678e-07, -5.2521e-08]
51117.8 Euler=[0.409095, 5.23826e-05, -0.409059]
 derivative=[-6.21396e-09, 5.02065e-07, -3.13045e-08]
51118 Euler=[0.409095, 5.25114e-05, -0.409059]
 derivative=[-6.21396e-09, 5.27218e-07, -7.71067e-09]
51118.2 Euler=[0.409095, 5.26456e-05, -0.409059]
 derivative=[-6.21396e-09, 5.45224e-07, 1.7913e-08]
51118.5 Euler=[0.409095, 5.27833e-05, -0.409059]
 derivative=[-6.21396e-09, 5.55277e-07, 4.51484e-08]
51118.8 Euler=[0.409095, 5.29225e-05, -0.409059]
 derivative=[-6.21396e-09, 5.56707e-07, 7.35139e-08]
51119 Euler=[0.409095, 5.30609e-05, -0.409059]
 derivative=[-6.21396e-09, 5.49021e-07, 1.02475e-07]
51119.2 Euler=[0.409095, 5.31962e-05, -0.409059]
 derivative=[-6.21396e-09, 5.31933e-07, 1.31456e-07]
51119.5 Euler=[0.409095, 5.33261e-05, -0.409058]
 derivative=[-6.21396e-09, 5.05387e-07, 1.59856e-07]
51119.8 Euler=[0.409095, 5.34481e-05, -0.409058]
 derivative=[-6.21396e-09, 4.69577e-07, 1.87066e-07]
51120 Euler=[0.409095, 5.35601e-05, -0.409058]
 derivative=[-6.21396e-09, 4.24955e-07, 2.12486e-07]
51120.2 Euler=[0.409095, 5.366e-05, -0.409058]
 derivative=[-6.21396e-09, 3.72226e-07, 2.35545e-07]
51120.5 Euler=[0.409095, 5.37457e-05, -0.409058]
 derivative=[-6.21396e-09, 3.1234e-07, 2.55718e-07]
51120.8 Euler=[0.409095, 5.38156e-05, -0.409058]
 derivative=[-6.21396e-09, 2.46465e-07, 2.72542e-07]
51121 Euler=[0.409095, 5.38685e-05, -0.409058]
 derivative=[-6.21396e-09, 1.75952e-07, 2.85634e-07]
51121.2 Euler=[0.409095, 5.39033e-05, -0.409058]
 derivative=[-6.21396e-09, 1.02297e-07, 2.94702e-07]
51121.5 Euler=[0.409095, 5.39195e-05, -0.409058]
 derivative=[-6.21396e-09, 2.70914e-08, 2.99554e-07]
51121.8 Euler=[0.409095, 5.39169e-05, -0.409058]
 derivative=[-6.21396e-09, -4.80341e-08, 3.00103e-07]
51122 Euler=[0.409095, 5.38956e-05, -0.409058]
 derivative=[-6.21396e-09, -1.21459e-07, 2.96373e-07]
51122.2 Euler=[0.409095, 5.38564e-05, -0.409058]
 derivative=[-6.21396e-09, -1.91629e-07, 2.88492e-07]
51122.5 Euler=[0.409095, 5.38002e-05, -0.409058]
 derivative=[-6.21396e-09, -2.57104e-07, 2.76691e-07]
51122.8 Euler=[0.409095, 5.37284e-05, -0.409058]
 derivative=[-6.21396e-09, -3.166e-07, 2.61289e-07]
51123 Euler=[0.409095, 5.36425e-05, -0.409058]
 derivative=[-6.21396e-09, -3.69027e-07, 2.42689e-07]
51123.2 Euler=[0.409095, 5.35445e-05, -0.409057]
 derivative=[-6.21396e-09, -4.13513e-07, 2.21356e-07]
51123.5 Euler=[0.409095, 5.34365e-05, -0.409057]
 derivative=[-6.21396e-09, -4.49421e-07, 1.97805e-07]
51123.8 Euler=[0.409095, 5.33206e-05, -0.409057]
 derivative=[-6.21396e-09, -4.76352e-07, 1.72583e-07]

 *** tNutation_tmp.out_a1
51116.2 Euler=[0.409095, 5.18002e-05, -0.409058]
 derivative=[-6.21396e-09, 2.54889e-07, -1.15973e-07]
51116.5 Euler=[0.409095, 5.18699e-05, -0.409058]
 derivative=[-6.21396e-09, 3.02605e-07, -1.0927e-07]
51116.8 Euler=[0.409095, 5.19514e-05, -0.409059]
 derivative=[-6.21396e-09, 3.48999e-07, -9.95502e-08]
51117 Euler=[0.409095, 5.20442e-05, -0.409059]
 derivative=[-6.21396e-09, 3.93136e-07, -8.68045e-08]
51117.2 Euler=[0.409095, 5.21477e-05, -0.409059]
 derivative=[-6.21396e-09, 4.34032e-07, -7.10869e-08]
51117.5 Euler=[0.409095, 5.22609e-05, -0.409059]
 derivative=[-6.21396e-09, 4.70678e-07, -5.2521e-08]
51117.8 Euler=[0.409095, 5.23826e-05, -0.409059]
 derivative=[-6.21396e-09, 5.02065e-07, -3.13045e-08]
51118 Euler=[0.409095, 5.25114e-05, -0.409059]
 derivative=[-6.21396e-09, 5.27218e-07, -7.71067e-09]
51118.2 Euler=[0.409095, 5.26456e-05, -0.409059]
 derivative=[-6.21396e-09, 5.45224e-07, 1.7913e-08]
51118.5 Euler=[0.409095, 5.27833e-05, -0.409059]
 derivative=[-6.21396e-09, 5.55277e-07, 4.51484e-08]
51118.8 Euler=[0.409095, 5.29225e-05, -0.409059]
 derivative=[-6.21396e-09, 5.56707e-07, 7.35139e-08]
51119 Euler=[0.409095, 5.30609e-05, -0.409059]
 derivative=[-6.21396e-09, 5.49021e-07, 1.02475e-07]
51119.2 Euler=[0.409095, 5.31962e-05, -0.409059]
 derivative=[-6.21396e-09, 5.31933e-07, 1.31456e-07]
51119.5 Euler=[0.409095, 5.33261e-05, -0.409058]
 derivative=[-6.21396e-09, 5.05387e-07, 1.59856e-07]
51119.8 Euler=[0.409095, 5.34481e-05, -0.409058]
 derivative=[-6.21396e-09, 4.69577e-07, 1.87066e-07]
51120 Euler=[0.409095, 5.35601e-05, -0.409058]
 derivative=[-6.21396e-09, 4.24955e-07, 2.12486e-07]
51120.2 Euler=[0.409095, 5.366e-05, -0.409058]
 derivative=[-6.21396e-09, 3.72226e-07, 2.35545e-07]
51120.5 Euler=[0.409095, 5.37457e-05, -0.409058]
 derivative=[-6.21396e-09, 3.1234e-07, 2.55718e-07]
51120.8 Euler=[0.409095, 5.38156e-05, -0.409058]
 derivative=[-6.21396e-09, 2.46465e-07, 2.72542e-07]
51121 Euler=[0.409095, 5.38685e-05, -0.409058]
 derivative=[-6.21396e-09, 1.75952e-07, 2.85634e-07]
51121.2 Euler=[0.409095, 5.39033e-05, -0.409058]
 derivative=[-6.21396e-09, 1.02297e-07, 2.94702e-07]
51121.5 Euler=[0.409095, 5.39195e-05, -0.409058]
 derivative=[-6.21396e-09, 2.70914e-08, 2.99554e-07]
51121.8 Euler=[0.409095, 5.39169e-05, -0.409058]
 derivative=[-6.21396e-09, -4.80341e-08, 3.00103e-07]
51122 Euler=[0.409095, 5.38956e-05, -0.409058]
 derivative=[-6.21396e-09, -1.21459e-07, 2.96373e-07]
51122.2 Euler=[0.409095, 5.38564e-05, -0.409058]
 derivative=[-6.21396e-09, -1.91629e-07, 2.88492e-07]
51122.5 Euler=[0.409095, 5.38002e-05, -0.409058]
 derivative=[-6.21396e-09, -2.57104e-07, 2.76691e-07]
51122.8 Euler=[0.409095, 5.37284e-05, -0.409058]
 derivative=[-6.21396e-09, -3.166e-07, 2.61289e-07]
51123 Euler=[0.409095, 5.36425e-05, -0.409058]
 derivative=[-6.21396e-09, -3.69027e-07, 2.42689e-07]
51123.2 Euler=[0.409095, 5.35445e-05, -0.409057]
 derivative=[-6.21396e-09, -4.13513e-07, 2.21356e-07]
51123.5 Euler=[0.409095, 5.34365e-05, -0.409057]
 derivative=[-6.21396e-09, -4.49421e-07, 1.97805e-07]
51123.8 Euler=[0.409095, 5.33206e-05, -0.409057]
 derivative=[-6.21396e-09, -4.76352e-07, 1.72583e-07]
51124 Euler=[0.409095, 5.31991e-05, -0.409057]
 derivative=[-6.21396e-09, -4.94142e-07, 1.4625e-07]

 *** tNutation_tmp.out_a2
51116.5 Euler=[0.409095, 5.18699e-05, -0.409058]
 derivative=[-6.21396e-09, 3.02605e-07, -1.0927e-07]
51116.8 Euler=[0.409095, 5.19514e-05, -0.409059]
 derivative=[-6.21396e-09, 3.48999e-07, -9.95502e-08]
51117 Euler=[0.409095, 5.20442e-05, -0.409059]
 derivative=[-6.21396e-09, 3.93136e-07, -8.68045e-08]
51117.2 Euler=[0.409095, 5.21477e-05, -0.409059]
 derivative=[-6.21396e-09, 4.34032e-07, -7.10869e-08]
51117.5 Euler=[0.409095, 5.22609e-05, -0.409059]
 derivative=[-6.21396e-09, 4.70678e-07, -5.2521e-08]
51117.8 Euler=[0.409095, 5.23826e-05, -0.409059]
 derivative=[-6.21396e-09, 5.02065e-07, -3.13045e-08]
51118 Euler=[0.409095, 5.25114e-05, -0.409059]
 derivative=[-6.21396e-09, 5.27218e-07, -7.71067e-09]
51118.2 Euler=[0.409095, 5.26456e-05, -0.409059]
 derivative=[-6.21396e-09, 5.45224e-07, 1.7913e-08]
51118.5 Euler=[0.409095, 5.27833e-05, -0.409059]
 derivative=[-6.21396e-09, 5.55277e-07, 4.51484e-08]
51118.8 Euler=[0.409095, 5.29225e-05, -0.409059]
 derivative=[-6.21396e-09, 5.56707e-07, 7.35139e-08]
51119 Euler=[0.409095, 5.30609e-05, -0.409059]
 derivative=[-6.21396e-09, 5.49021e-07, 1.02475e-07]
51119.2 Euler=[0.409095, 5.31962e-05, -0.409059]
 derivative=[-6.21396e-09, 5.31933e-07, 1.31456e-07]
51119.5 Euler=[0.409095, 5.33261e-05, -0.409058]
 derivative=[-6.21396e-09, 5.05387e-07, 1.59856e-07]
51119.8 Euler=[0.409095, 5.34481e-05, -0.409058]
 derivative=[-6.21396e-09, 4.69577e-07, 1.87066e-07]
51120 Euler=[0.409095, 5.35601e-05, -0.409058]
 derivative=[-6.21396e-09, 4.24955e-07, 2.12486e-07]
51120.2 Euler=[0.409095, 5.366e-05, -0.409058]
 derivative=[-6.21396e-09, 3.72226e-07, 2.35545e-07]
51120.5 Euler=[0.409095, 5.37457e-05, -0.409058]
 derivative=[-6.21396e-09, 3.1234e-07, 2.55718e-07]
51120.8 Euler=[0.409095, 5.38156e-05, -0.409058]
 derivative=[-6.21396e-09, 2.46465e-07, 2.72542e-07]
51121 Euler=[0.409095, 5.38685e-05, -0.409058]
 derivative=[-6.21396e-09, 1.75952e-07, 2.85634e-07]
51121.2 Euler=[0.409095, 5.39033e-05, -0.409058]
 derivative=[-6.21396e-09, 1.02297e-07, 2.94702e-07]
51121.5 Euler=[0.409095, 5.39195e-05, -0.409058]
 derivative=[-6.21396e-09, 2.70914e-08, 2.99554e-07]
51121.8 Euler=[0.409095, 5.39169e-05, -0.409058]
 derivative=[-6.21396e-09, -4.80341e-08, 3.00103e-07]
51122 Euler=[0.409095, 5.38956e-05, -0.409058]
 derivative=[-6.21396e-09, -1.21459e-07, 2.96373e-07]
51122.2 Euler=[0.409095, 5.38564e-05, -0.409058]
 derivative=[-6.21396e-09, -1.91629e-07, 2.88492e-07]
51122.5 Euler=[0.409095, 5.38002e-05, -0.409058]
 derivative=[-6.21396e-09, -2.57104e-07, 2.76691e-07]
51122.8 Euler=[0.409095, 5.37284e-05, -0.409058]
 derivative=[-6.21396e-09, -3.166e-07, 2.61289e-07]
51123 Euler=[0.409095, 5.36425e-05, -0.409058]
 derivative=[-6.21396e-09, -3.69027e-07, 2.42689e-07]
51123.2 Euler=[0.409095, 5.35445e-05, -0.409057]
 derivative=[-6.21396e-09, -4.13513e-07, 2.21356e-07]
51123.5 Euler=[0.409095, 5.34365e-05, -0.409057]
 derivative=[-6.21396e-09, -4.49421e-07, 1.97805e-07]
51123.8 Euler=[0.409095, 5.33206e-05, -0.409057]
 derivative=[-6.21396e-09, -4.76352e-07, 1.72583e-07]
51124 Euler=[0.409095, 5.31991e-05, -0.409057]
 derivative=[-6.21396e-09, -4.94142e-07, 1.4625e-07]
51124.2 Euler=[0.409095, 5.30743e-05, -0.409057]
 derivative=[-6.21396e-09, -5.0285e-07, 1.19363e-07]

 *** tNutation_tmp.out_a3
51116.8 Euler=[0.409095, 5.19514e-05, -0.409059]
 derivative=[-6.21396e-09, 3.48999e-07, -9.95502e-08]
51117 Euler=[0.409095, 5.20442e-05, -0.409059]
 derivative=[-6.21396e-09, 3.93136e-07, -8.68045e-08]
51117.2 Euler=[0.409095, 5.21477e-05, -0.409059]
 derivative=[-6.21396e-09, 4.34032e-07, -7.10869e-08]
51117.5 Euler=[0.409095, 5.22609e-05, -0.409059]
 derivative=[-6.21396e-09, 4.70678e-07, -5.2521e-08]
51117.8 Euler=[0.409095, 5.23826e-05, -0.409059]
 derivative=[-6.21396e-09, 5.02065e-07, -3.13045e-08]
51118 Euler=[0.409095, 5.25114e-05, -0.409059]
 derivative=[-6.21396e-09, 5.27218e-07, -7.71067e-09]
51118.2 Euler=[0.409095, 5.26456e-05, -0.409059]
 derivative=[-6.21396e-09, 5.45224e-07, 1.7913e-08]
51118.5 Euler=[0.409095, 5.27833e-05, -0.409059]
 derivative=[-6.21396e-09, 5.55277e-07, 4.51484e-08]
51118.8 Euler=[0.409095, 5.29225e-05, -0.409059]
 derivative=[-6.21396e-09, 5.56707e-07, 7.35139e-08]
51119 Euler=[0.409095, 5.30609e-05, -0.409059]
 derivative=[-6.21396e-09, 5.49021e-07, 1.02475e-07]
51119.2 Euler=[0.409095, 5.31962e-05, -0.409059]
 derivative=[-6.21396e-09, 5.31933e-07, 1.31456e-07]
51119.5 Euler=[0.409095, 5.33261e-05, -0.409058]
 derivative=[-6.21396e-09, 5.05387e-07, 1.59856e-07]
51119.8 Euler=[0.409095, 5.34481e-05, -0.409058]
 derivative=[-6.21396e-09, 4.69577e-07, 1.87066e-07]
51120 Euler=[0.409095, 5.35601e-05, -0.409058]
 derivative=[-6.21396e-09, 4.24955e-07, 2.12486e-07]
51120.2 Euler=[0.409095, 5.366e-05, -0.409058]
 derivative=[-6.21396e-09, 3.72226e-07, 2.35545e-07]
51120.5 Euler=[0.409095, 5.37457e-05, -0.409058]
 derivative=[-6.21396e-09, 3.1234e-07, 2.55718e-07]
51120.8 Euler=[0.409095, 5.38156e-05, -0.409058]
 derivative=[-6.21396e-09, 2.46465e-07, 2.72542e-07]
51121 Euler=[0.409095, 5.38685e-05, -0.409058]
 derivative=[-6.21396e-09, 1.75952e-07, 2.85634e-07]
51121.2 Euler=[0.409095, 5.39033e-05, -0.409058]
 derivative=[-6.21396e-09, 1.02297e-07, 2.94702e-07]
51121.5 Euler=[0.409095, 5.39195e-05, -0.409058]
 derivative=[-6.21396e-09, 2.70914e-08, 2.99554e-07]
51121.8 Euler=[0.409095, 5.39169e-05, -0.409058]
 derivative=[-6.21396e-09, -4.80341e-08, 3.00103e-07]
51122 Euler=[0.409095, 5.38956e-05, -0.409058]
 derivative=[-6.21396e-09, -1.21459e-07, 2.96373e-07]
51122.2 Euler=[0.409095, 5.38564e-05, -0.409058]
 derivative=[-6.21396e-09, -1.91629e-07, 2.88492e-07]
51122.5 Euler=[0.409095, 5.38002e-05, -0.409058]
 derivative=[-6.21396e-09, -2.57104e-07, 2.76691e-07]
51122.8 Euler=[0.409095, 5.37284e-05, -0.409058]
 derivative=[-6.21396e-09, -3.166e-07, 2.61289e-07]
51123 Euler=[0.409095, 5.36425e-05, -0.409058]
 derivative=[-6.21396e-09, -3.69027e-07, 2.42689e-07]
51123.2 Euler=[0.409095, 5.35445e-05, -0.409057]
 derivative=[-6.21396e-09, -4.13513e-07, 2.21356e-07]
51123.5 Euler=[0.409095, 5.34365e-05, -0.409057]
 derivative=[-6.21396e-09, -4.49421e-07, 1.97805e-07]
51123.8 Euler=[0.409095, 5.33206e-05, -0.409057]
 derivative=[-6.21396e-09, -4.76352e-07, 1.72583e-07]
51124 Euler=[0.409095, 5.31991e-05, -0.409057]
 derivative=[-6.21396e-09, -4.94142e-07, 1.4625e-07]
51124.2 Euler=[0.409095, 5.30743e-05, -0.409057]
 derivative=[-6.21396e-09, -5.0285e-07, 1.19363e-07]
51124.5 Euler=[0.409095, 5.29484e-05, -0.409057]
 derivative=[-6.21396e-09, -5.02734e-07, 9.24611e-08]
